service/common/src/main/java/org/apache/polaris/service/catalog/conversion/xtable/XTableConverter.java

```diff
@@ -0,0 +1,116 @@
+  private static final String RUN_SYNC_ENDPOINT = "/v1/conversion/sync";
+
+  private final String hostUrl;
+  private final HttpClient client;
+  private final ObjectMapper objectMapper;
+
+  public XTableConverter(String hostUrl) {
+    this.hostUrl = hostUrl;
+    this.client =
+        HttpClient.newBuilder()
+            .version(HttpClient.Version.HTTP_2)
+            .followRedirects(HttpClient.Redirect.NORMAL)
+            .build();
+    this.objectMapper = new ObjectMapper();
+  }
+
+  @VisibleForTesting
+  XTableConverter(String hostUrl, HttpClient client, ObjectMapper objectMapper) {
+    this.hostUrl = hostUrl;
+    this.client = client;
+    this.objectMapper = objectMapper;
+  }
+
+  public RunSyncResponse execute(TableLikeEntity tableEntity) {
+    String sourceFormat;
+    String sourceMetadataPath;
+    String targetFormat;
+
+    switch (tableEntity.getSubType()) {
+      case GENERIC_TABLE -> {
+        GenericTableEntity generic = (GenericTableEntity) tableEntity;
+        sourceFormat = checkIfSupportedFormat(generic.getFormat());
+        sourceMetadataPath = generic.getPropertiesAsMap().get(SOURCE_METADATA_PATH_KEY);
+        targetFormat =
+            checkIfSupportedFormat(
+                generic.getPropertiesAsMap().get(ENABLED_READ_TABLE_FORMATS_KEY));
+      }
+      case ICEBERG_TABLE -> {
+        IcebergTableLikeEntity iceberg = (IcebergTableLikeEntity) tableEntity;
+        sourceFormat = TableFormat.ICEBERG.name();
+        sourceMetadataPath = iceberg.getMetadataLocation();
+        targetFormat = iceberg.getPropertiesAsMap().get(ENABLED_READ_TABLE_FORMATS_KEY);
+      }
+      default ->
+          throw new IllegalArgumentException(
+              "Unsupported TableEntity type: " + tableEntity.getSubType());
+    }
+
+    return executeRunSyncRequest(sourceFormat, sourceMetadataPath, targetFormat);
+  }
+
+  private RunSyncResponse executeRunSyncRequest(
+      String sourceFormat, String sourceMetadataPath, String targetFormat) {
+
+    RunSyncRequest request = new RunSyncRequest(sourceFormat, sourceMetadataPath, targetFormat);
+    try {
+      String requestBody = objectMapper.writeValueAsString(request);
+      HttpRequest httpRequest =
+          HttpRequest.newBuilder()
+              .uri(URI.create(hostUrl + RUN_SYNC_ENDPOINT))
+              .header("Accept", "application/json")
+              .POST(HttpRequest.BodyPublishers.ofString(requestBody))
+              .build();
+
+      HttpResponse<String> response =
+          client.send(httpRequest, HttpResponse.BodyHandlers.ofString());
+      if (response.statusCode() != SUCCESS_STATUS_CODE) {
+        throw new IllegalStateException("Conversion failed: " + response.body());
+      }
+      return objectMapper.readValue(response.body(), RunSyncResponse.class);
+    } catch (IOException | InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static String checkIfSupportedFormat(String format) {
```

Comment:
Do we want to do this on the client side or simply let the xtable server return an error? XTable has the ability for users to register formats that are not in open source

Reply:
Yea i think i wanted to save the extra network hop, since its simple for the client to validate this.